### PR TITLE
Bach: POC graphing SQLModels

### PR DESCRIPTION
**Pull request by KathiaBarahona**
Some obscure idea I was exploring last time when analyzing queries, played around with a couple of graphing packages and `graphviz` was the best.
I'm still missing unit tests
Here an example of what I managed to do:

For:
```python
from modelhub import ModelHub
modelhub = ModelHub(time_aggregation='%Y-%m-%d')
df = modelhub.get_objectiv_dataframe(start_date='2021-11-16')
df['is_new_user'] = modelhub.map.is_new_user(df)

modelhub.add_conversion_event(
    location_stack=df.location_stack.json[{'id': 'Quickstart Guide', '_type': 'LinkContext'}:],
    event_type='PressEvent',
    name='quickstart_presses'
)
df['conversion_events'] = modelhub.map.is_conversion_event(df, 'quickstart_presses')
df['is_new_user_month'] = modelhub.map.is_new_user(df, time_aggregation='%Y-%m')
df['is_twice_converted'] = modelhub.map.conversions_in_time(df, name='quickstart_presses') == 2

unique_users = modelhub.aggregate.unique_users(df[df.is_new_user_month & df.is_twice_converted])

from bach.display_formats import display_sql_as_graph
display_sql_as_graph(unique_users)
```
image

### CODE REVIEW

File: bach/bach/display_formats/graph_display.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+def display_sql_as_graph(obj: DataFrameOrSeries) -> None:
+    """
+    Displays a graph of all referenced nodes by :py:meth:`DataFrame.base_node` in the frontend (notebook)
+    as an SVG.
+
+    :param obj: DataFrame or Series from where to obtain the generated sql.
+
+    .. note::
+        Requires the graphviz package, if not installed the query will be print instead
```
Comment (vard-obj) [inline bach/bach/display_formats/graph_display.py:203]: I'll get `AttributeError: 'SqlModelGraphDisplay' object has no attribute '_graph'` if I don't install graphviz in the system.

**Review by thijs-obj — COMMENTED** [review]
Had a first look. Think this is a very cool feature.

Not sure how to review this. It says POC and that tests should still be added, but should I consider other code to be review ready?

One general comment at least is that sometimes the graph renders very wide. There is probably something to force graphviz to render more vertical, which might give nicer graphs in some cases.
An example of a graph that's more horizontal than vertical is in the basic-product-analytics notebook. I tried replacing the describe call in the 7th cell with this: `display_sql_as_graph(df.describe(include='all'))`. I believe the rendered graph is correct, but it is kinda hard to grok as it is so wide.

**Comment by KathiaBarahona** [comment]
> Had a first look. Think this is a very cool feature.
> 
> Not sure how to review this. It says POC and that tests should still be added, but should I consider other code to be review ready?
> 
> One general comment at least is that sometimes the graph renders very wide. There is probably something to force graphviz to render more vertical, which might give nicer graphs in some cases. An example of a graph that's more horizontal than vertical is in the basic-product-analytics notebook. I tried replacing the describe call in the 7th cell with this: `display_sql_as_graph(df.describe(include='all'))`. I believe the rendered graph is correct, but it is kinda hard to grok as it is so wide.

As discussed in planning, I added a warning in the docstring that this more of an experimental functionality

File: bach/mypy.ini [hunk 2]
```diff
@@ -22,3 +22,6 @@ ignore_missing_imports = True
 
 [mypy-sqlalchemy_bigquery]
 ignore_missing_imports = True
+
+[mypy-graphviz.*]
+ignore_missing_imports = True
```
Comment (jansentom) [inline bach/mypy.ini:27]:
```suggestion
ignore_missing_imports = True

```

File: bach/bach/display_formats/graph_display.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+    .. note::
+        Requires the IPython package, if not installed the graph source will be print instead
```
Comment (jansentom) [inline bach/bach/display_formats/graph_display.py:213]:
```suggestion
    .. note::
        Requires the graphviz and IPython package, if not installed the query will be printed instead.
```

File: bach/bach/display_formats/graph_display.py [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+
+
+class SqlModelGraphDisplay:
+    def __init__(
+        self,
+        obj: DataFrameOrSeries,
+    ) -> None:
+        try:
+            from graphviz import Digraph
+            self._graph = Digraph(node_attr=DEFAULT_GRAPH_NODE_FORMAT)
```
Comment (jansentom) [inline bach/bach/display_formats/graph_display.py:31]: should `node_attr` not be a parameter? Like this it's not defaulted to `DEFAULT_GRAPH_NODE_FORMAT`, it's the only option, but maybe [that's okay](http://oplaunch.com/blog/2015/04/30/the-truth-about-any-color-so-long-as-it-is-black/)

**Review by jansentom — APPROVED** [review]
small remarks/suggestions, but LGTM otherwise. Thanks.

File: bach/bach/display_formats/graph_display.py (hunk 4 above)
Comment (KathiaBarahona) [reply bach/bach/display_formats/graph_display.py:31]: true true

Comment (KathiaBarahona) [reply bach/bach/display_formats/graph_display.py]: added here https://github.com/objectiv/objectiv-analytics/pull/986/commits/0e4cbad61d80fcebaee68df7ed988661e9fcf667

File: bach/bach/display_formats/graph_display.py (hunk 3 above)
Comment (KathiaBarahona) [reply bach/bach/display_formats/graph_display.py]: Done here https://github.com/objectiv/objectiv-analytics/pull/986/commits/0e4cbad61d80fcebaee68df7ed988661e9fcf667

File: bach/mypy.ini (hunk 2 above)
Comment (KathiaBarahona) [reply bach/mypy.ini]: https://github.com/objectiv/objectiv-analytics/pull/986/commits/0e4cbad61d80fcebaee68df7ed988661e9fcf667

**Pull request merged**